### ISSUE TITLE
[Improvement] Mark app data lost when encountering events dropped

### ISSUE THREAD

**Issue by zuston** [apache/incubator-uniffle#372]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

When shuffle-server encounters disk problems and triggers the mechanism of dropping events, we should mark the app data lost.

After then, we could reject the write/read requests by these apps to make job fast fail. By the way, this mechanism is compatible with the multiple replica

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

**Comment by zuston** [comment]
@xianjingfeng @jerqi PTAL. I think this is an improvement.

**Comment by jerqi** [comment]
> @xianjingfeng @jerqi PTAL. I think this is an improvement.
 
For multi replicas, it won't be suitable. Because one replica don't need store complete data for multi replicas.

**Comment by zuston** [comment]
> > @xianjingfeng @jerqi PTAL. I think this is an improvement.
> 
> For multi replicas, it won't be suitable. Because one replica don't need store complete data for multi replicas.

I missed this point. So maybe we need to make job fast fail only when single replica is enabled?

**Comment by jerqi** [comment]
> > > @xianjingfeng @jerqi PTAL. I think this is an improvement.
> > 
> > 
> > For multi replicas, it won't be suitable. Because one replica don't need store complete data for multi replicas.
> 
> I missed this point. So maybe we need to make job fast fail only when single replica is enabled?

Server don't have replica concept.

**Comment by zuston** [comment]
We could populate this replica concept into ShuffleTaskInfo when client registers to shuffle-server.

**Comment by jerqi** [comment]
We should be careful about involving the replica concept in server. Especially, we need modify rpc proto. I think this will be an important concept. How to development the concept in the future? This is what we need to think of?
It's better to ask @xianjingfeng for some advice. Because he has more business needs about production environment.

**Comment by zuston** [comment]
Feel free to discuss more. Now we also suffer from the problem of single partition data too large and need this to make job fast fail to improve stability.

**Comment by jerqi** [comment]
> Feel free to discuss more. Now we also suffer from the problem of single partition data too large and need this to make job fast fail to improve stability.

For huge shuffle, we really recommend to use MEMORY_LOCALFILE_HDFS. You can use a large value `rss.server.flush.cold.storage.threshold.size` to reduce the data size of shuffle which were written to HDFS.

**Comment by zuston** [comment]
> > Feel free to discuss more. Now we also suffer from the problem of single partition data too large and need this to make job fast fail to improve stability.
> 
> For huge shuffle, we really recommend to use MEMORY_LOCALFILE_HDFS. You can use a large value `rss.server.flush.cold.storage.threshold.size` to reduce the data size of shuffle which were written to HDFS.

Yes. I have to evaluate this proposal of `MEMORY_LOCALFILE_HDFS`

**Comment by jerqi** [comment]
> Anyway, this issue is still reasonable.
Yes. you're right.

**Comment by xianjingfeng** [comment]
I think it is possible that both disk and hdfs can't be write. Maybe because of the network or GC. I am ok if just for one replica.
